crates/schema/src/def.rs

```diff
@@ -0,0 +1,640 @@
+//! Canonicalized module definitions.
+//!
+//! This module contains a set of types that represent the canonical form of SpacetimeDB module definitions.
+//! These types are immutable to prevent accidental introduction of errors.
+//! The internal data structures of this module are not considered public API or ABI and may change
+//! at any time.
+//!
+//! Different module ABI versions correspond to different submodules of `spacetimedb_lib::db::raw_def`.
+//! All of these ABI versions can be converted to the standard form in this module via `TryFrom`.
+//! We provide streams of errors in case the conversion fails, to provide as much information
+//! to the user as possible about why their module is invalid.
+//!
+//! The `ModuleDef` type is the main type in this module. It contains all the information about a module, including its tables, reducers, typespace, and type metadata.
+//!
+//! After validation, a `ModuleDef` can be converted to the `*Schema` types in `crate::schema` for use in the database.
+//! (Eventually, we may unify these types...)
+
+use std::fmt::{self, Write};
+
+use crate::error::{IdentifierError, ValidationErrors};
+use crate::identifier::Identifier;
+use itertools::Itertools;
+use spacetimedb_data_structures::error_stream::{CollectAllErrors, CombineErrors, ErrorStream};
+use spacetimedb_data_structures::map::HashMap;
+use spacetimedb_lib::db::auth::{StAccess, StTableType};
+use spacetimedb_lib::db::raw_def;
+use spacetimedb_lib::db::raw_def::v9::{RawIdentifier, RawScopedTypeNameV9};
+use spacetimedb_lib::{ProductType, RawModuleDef};
+use spacetimedb_primitives::{ColId, ColList};
+use spacetimedb_sats::AlgebraicType;
+use spacetimedb_sats::{AlgebraicTypeRef, Typespace};
+
+pub mod validate;
+
+/// A map from `Identifier`s to values of type `T`.
+pub type IdentifierMap<T> = HashMap<Identifier, T>;
```

Comment:
As far as I can tell, all the types you store here implement your nice `ModuleDefLookup` trait and already allow to get a map key from `T`.

One thing I've done in the past for similar scenarios is use `HashSet` instead + a newtype that implements `Eq + Hash + Borrow<Key>` by delegating to the key comparisons.

That way you should be able to avoid a lot of allocations in all these `(foobar.name.clone(), foobar)` KV construction lines, as well as save some memory by simply reusing `foobar` itself both for the key and the value parts of each entry.

Idk how critical it is, so feel free to leave this as a future optimisation, just thought I'd throw it out there as I noticed a lot of those clones in the PR.

Reply:
Oh that's a great idea. That should get rid of almost all of these clones and get rid of some possible invalid states.

I'm tempted to leave it as a future optimization for now though just in favor of getting this merged.